maven_iuvs/echelle.py

```diff
@@ -0,0 +1,886 @@
+            elif date[1] == -1: # Use this to just go until the present time/date.
+                date[1] = datetime.datetime.utcnow()
+            
+            print(f"Returning observations between {date[0]} and {date[1]}")
+
+            selected_lights = [entry for entry in selected_lights if date[0] <= entry['datetime'] <= date[1]]
+
+        # To get observations at a specific day or specific day/time:
+        elif type(date) is not list:  
+
+            if type(date) == datetime.date: # If no time information was entered, be liberal and assume start of first day and end of last
+                date0 = datetime.datetime(date.year, date.month, date.day, 0, 0, 0)
+                date1 = datetime.datetime(date.year, date.month, date.day, 23, 59, 59)
+
+                selected_lights = [entry for entry in selected_lights if date0 <= entry['datetime'] <= date1]
+
+            else: # if a full datetime.datetime object is entered, look for that exact entry.
+                selected_lights = [entry for entry in selected_lights if entry['datetime'] == date]
+
+        else:
+            raise TypeError(f"Date entered is of type {type(date)}")
+
+    return selected_lights
+
+# Relating to dark vs. light observations -----------------------------
+
+
+def coadd_lights(light_fits, dark_fits):
+    """
+    Co-add all light frames within light_fits, including subtraction
+    of dark frames in dark_fits.
+
+    Parameters
+    ----------
+    light_fits : IUVSFits object or HDUlist
+                 fits object representing the light observations.
+    dark_fits : IUVSFits object or HDUlist
+                 fits object representing the associated darks.
+
+    Returns
+    ----------
+    coadded_lights : array
+                     Essentially the mean-frame of the detector
+    nan_light_inds : array
+                     Indices of light frames in which there are nan values.
+    bad_light_inds : array
+                     Indices of light frames in which there are bad values (not nans)
+    nan_dark_inds : array
+                    Indices of dark frames in which there are nans.
+    total_frames : int
+                   total number of good frames which were used to create the 
+                   co-added image. Returned so the value can be added to 
+                   quicklooks.
+    """
+
+    # Retrieve dark frames for subtraction
+    first_dark, second_dark = get_dark_frames(dark_fits)
+
+    # check darks are valid
+    if np.isnan(first_dark).any() & np.isnan(second_dark).any():
+        raise Exception(f"Missing critical observation data: no valid darks")
+
+    # dark subtraction
+    dark_subtracted, nan_light_inds, bad_light_inds, nan_dark_inds = subtract_darks(light_fits, first_dark, second_dark)
+    
+    # check that we still have valid data
+    if dark_subtracted.shape[0] == 0: # Some files may have no acceptable frames
```

Comment:
exceptions on returned data should be raised inside the function that returns the data

Reply:
Fixed by 1fd9de2318dba503f971986065f883b8a930c043